### PR TITLE
[Fixes: #10746] Setup rate limits for all_auth

### PR DESCRIPTION
**Pull request by t-book**
Rate limits for all_auth. see #10746 

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/GeoNode/geonode/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in the documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for contrib modules):
- [x] There is a ticket in https://github.com/GeoNode/geonode/issues describing the issue/improvement/feature (a notable exemption is, changes not visible to end-users)
- [x] The issue connected to the PR must have Labels and Milestone assigned
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[Fixes #<issue_number>] Title of the Issue"
- [ ] New unit tests have been added covering the changes, unless there is an explanation on why the tests are not necessary/implemented
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the QA checks: black geonode && flake8 geonode
- [ ] Commits changing the **settings**, **UI**, **existing user workflows**, or adding **new functionality**, need to include documentation updates
- [ ] Commits adding **new texts** do use gettext and have updated .po / .mo files (without location infos)

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**

### CODE REVIEW

File: geonode/settings.py [hunk 1]
```diff
@@ -334,7 +334,9 @@
 CACHES = {
     # DUMMY CACHE FOR DEVELOPMENT
     "default": {
-        "BACKEND": "django.core.cache.backends.dummy.DummyCache",
```
Comment (afabiani) [inline geonode/settings.py:337]: I guess you removed the `BACKEND` key here erroneously

**Comment by giohappy** [comment]
@t-book default rate limits should be already provided by django_allauth (see my comment https://github.com/GeoNode/geonode/issues/10746#issuecomment-1460154081).

@afabiani @t-book the changing the cache backend is quite a critical component. Are we sure we should do it so light-hearted? I think it will require extensive testing.

File: geonode/settings.py (hunk 1 above)
Comment (t-book) [reply geonode/settings.py:337]: Oh yes! Thanks @afabiani

**Comment by t-book** [comment]
@giohappy Yes definitly. This change should be tested carefully. Maybe let's keep it as draft:

> default rate limits should be already provided by django_allauth 

https://github.com/pennersr/django-allauth/blob/7c50e111078e481aa65804f88f2366a3c013b909/allauth/account/app_settings.py#L175-L191

From what I know this will not work out of the box:

> In order to be secure out of the box various rate limits are in place. The rate limit mechanism is backed by a Django cache. **Hence, rate limiting will not work properly if you are using the DummyCache.**

https://django-allauth.readthedocs.io/en/latest/configuration.html?highlight=cache#configuration

**Comment by afabiani** [comment]
> @t-book default rate limits should be already provided by django_allauth (see my comment [#10746 (comment)](https://github.com/GeoNode/geonode/issues/10746#issuecomment-1460154081)).
> 
> @afabiani @t-book that changing the cache backend is quite an important change. Are we sure we should do it so light-hearted? I think it will require extensive testing.

Agree, it is also not recommended to use `LocMemCache` in a production system (as far as I know)

**Comment by t-book** [comment]
> Agree, it is also not recommended to use LocMemCache in a production system (as far as I know)

The downsides I know of:

- It is not very memory efficient
- as it stores data in memory, data will be lost on a server restart 
- and it is not scalable across multiple servers

On the other hand I'm unsure if adding Redis or Memcached to the stack is really needed for this task. 
For sure it is good to think about / discuss it.

**Comment by t-book** [comment]
I've removed the RATE_LIMITS from settings.py as default values are fine.
Further I will also test this with LDAP enabled ...

**Comment by t-book** [comment]
@giohappy I've seen we already install memcached
https://github.com/GeoNode/geonode/blob/a4a21b72b0ceb6ebf60d397078d0b42892869396/Dockerfile#L45
So this means we could do the following to setup the rate limits:
- Enable memcached for production which allows all_auth to use the rate limits (caching backend needs to be tested also contrib app like risks which looks like using fe caching)
- add the error 429.html template provided by this PR

**Comment by giohappy** [comment]
@t-book AFAIR caching was disabled because it was giving problems with DRF. Before re-enabling it we would need to test the side effects thoroughly.

You know, caching with such a dynamic API (based on current user, current permissions, etc.) is not an easy task. It will have to be managed view by view, taking care to not make the cache kick in in places where it shouldn't.

**Comment by t-book** [comment]
@giohappy you are absolutely right! This step should be treated with caution. Sadly all_auth relies on caching for its limits. Another option might be creating a custom middleware that blocks DOS attacks for mailing endpoints. I will think about it.

**Comment by t-book** [comment]
In case the caching is no option we could add a small app to contrib_apps that prevents email flooting and relies on the DB. Untested but to show the idea ...

model
```python
from django.db import models

class RequestTracker(models.Model):
    ip_address = models.CharField(max_length=255)
    url = models.CharField(max_length=255)
    timestamp = models.DateTimeField(auto_now_add=True)
```

middleware
```python
from django.core.exceptions import PermissionDenied
from django.utils import timezone
from .models import RequestTracker

class RequestLimiterMiddleware:
    def __init__(self, get_response):
        self.get_response = get_response

    def __call__(self, request):
        if request.method == 'POST' and request.path in ['/account/signup/', '/account/password/reset/', '/account/login/']:
            ip_address = request.META.get('HTTP_X_FORWARDED_FOR') or request.META.get('REMOTE_ADDR')
            current_time = timezone.now()
            one_minute_ago = current_time - timezone.timedelta(minutes=1)
            num_requests = RequestTracker.objects.filter(ip_address=ip_address, url=request.path, timestamp__gte=one_minute_ago).count()
            if num_requests >= 5:
                raise PermissionDenied('Too many requests in one minute')
            RequestTracker.objects.create(ip_address=ip_address, url=request.path)
        response = self.get_response(request)
        return response
```

**Pull request closed without merging**